velox/vector/ComplexVector.cpp

```diff
@@ -1781,4 +1782,266 @@ void RowVector::appendNulls(vector_size_t numberOfRows) {
   bits::fillBits(mutableRawNulls(), oldSize, newSize, bits::kNull);
 }
 
+namespace {
+template <typename KeyType, TypeKind ValueKind>
+std::vector<VectorPtr> toRowVectorComplexImpl(
+    const BaseVector& vector,
+    folly::F14FastMap<KeyType, size_t> keyToIndex,
+    bool replaceNulls,
+    const SelectivityVector& rows) {
+  if constexpr (ValueKind != TypeKind::ARRAY && ValueKind != TypeKind::MAP) {
+    if (replaceNulls) {
+      VELOX_FAIL(
+          "Unsupported type for replacing nulls: {}",
+          TypeTraits<ValueKind>::TypeTraits::name);
+    }
+  }
+  DecodedVector decodedVector(vector);
+  auto mapBase = decodedVector.base()->asUnchecked<MapVector>();
+  auto valueType = mapBase->mapValues()->type();
+  auto mapKeys = mapBase->mapKeys();
+  auto decodedKeys = DecodedVector(*mapKeys);
+  auto mapValues = mapBase->mapValues();
+  auto decodedValues = DecodedVector(*mapValues);
+  auto offsets = mapBase->rawOffsets();
+  auto sizes = mapBase->rawSizes();
+  vector_size_t outputSize = rows.end();
+  std::vector<bool> visited(keyToIndex.size());
+
+  std::vector<VectorPtr> children;
+  children.reserve(keyToIndex.size());
+  std::vector<std::vector<BaseVector::CopyRange>> copyRangesFromBase(
+      keyToIndex.size());
+  for (size_t i = 0; i < keyToIndex.size(); ++i) {
+    children.push_back(
+        BaseVector::create(valueType, outputSize, mapBase->pool()));
+  }
+
+  rows.applyToSelected([&](vector_size_t row) INLINE_LAMBDA {
+    std::fill(visited.begin(), visited.end(), false);
+    if (!decodedVector.isNullAt(row)) {
+      auto decodedIndex = decodedVector.index(row);
+      auto offset = offsets[decodedIndex];
+      auto size = sizes[decodedIndex];
+      for (auto j = offset; j < offset + size; ++j) {
+        if (decodedKeys.isNullAt(j) || decodedValues.isNullAt(j)) {
+          continue;
+        }
+        auto key = decodedKeys.valueAt<KeyType>(j);
+        auto it = keyToIndex.find(key);
+        if (it == keyToIndex.end()) {
+          continue;
+        }
+        auto index = it->second;
+        visited[index] = true;
```

Comment:
Should we check that visited[index] is not already true (e.g. fail if there are duplicate map keys)?